File: languages/rust/rust_vs_c/stack_values/c/array.c
#include <stdio.h>

typedef unsigned char u8;

void print_arr(u8 arr[], int len)
{
  putchar('[');
  for (int i = 0; i < len; i++)
  {
    printf("%d,", arr[i]);
  }
  puts("]");
}

// void mod_by_ptr(u8 arr[])
void mod_by_ptr(u8 *arr)
{
  arr[0] = 99;
}

typedef struct
{
  u8 arr[3];
} ArrayContainer;

void mod_by_copy(ArrayContainer arr_container)
{
  arr_container.arr[0] = 99;
}

int main()
{
  puts("---mod_by_ptr---");
  u8 arr[] = {1, 2, 3};
  print_arr(arr, 3);
  mod_by_ptr(arr);
  print_arr(arr, 3);

  puts("---mod_by_copy---");
  ArrayContainer arr_container = {.arr = {1, 2, 3}};
  print_arr(arr_container.arr, 3);
  mod_by_copy(arr_container);
  print_arr(arr_container.arr, 3);
}

File: languages/rust/rust_vs_c/stack_values/c/array.js
/**
 * @param {number[]} arr (C - number *) (Rust - &mut [i32])
 */
function mod_by_ptr(arr) {
  arr[0] = 99;
}

/** @typedef { { arr: number[] } } Ctnr */

/**
 * @param {Ctnr} ctnr (C - Ctnr *) (Rust - &mut Ctnr)
 */
function mod_by_another_pointer(ctnr) {
  // ctnr = {};
  ctnr.arr[0] = 99;
}

console.log("---mod_by_ptr---");
const arr = [1, 2, 3];
console.log(arr);
mod_by_ptr(arr); // (&arr)
console.log(arr);

console.log("---mod_by_another_ptr---");
const arr_temp = [1, 2, 3];
const arr_container = { arr: arr_temp };
console.log(arr_temp);
mod_by_another_pointer(arr_container); // (&arr_container)
console.log(arr_temp);
